# Exercise 1 — Potter Airlines

**Instructor solution: about 70% is provided and about 30% is shown with completed solutions.**

Run the cells in order. Complete the sections marked **TODO**. The goal is to practice dictionaries, lists, loops, conditions, and functions without starting from a blank notebook.


## Part A — Flight data


In [3]:
flight = {
    "flight_id": "PA1001",
    "origin": "London",
    "destination": "Hogsmeade",
    "departure_date": "2026-09-26",
    "days_until_departure": 12,
    "base_fare": 135.0,
    "seats_remaining": 18,
    "capacity": 120,
    "route_popularity": 0.91,
}

flights = [
    flight,
    {"flight_id": "PA1002", "origin": "London", "destination": "Edinburgh",
     "departure_date": "2026-10-19", "days_until_departure": 35,
     "base_fare": 92.0, "seats_remaining": 70, "capacity": 120, "route_popularity": 0.72},
    {"flight_id": "PA1003", "origin": "Manchester", "destination": "Diagon Alley",
     "departure_date": "2026-09-17", "days_until_departure": 3,
     "base_fare": 96.0, "seats_remaining": 8, "capacity": 90, "route_popularity": 0.69},
    {"flight_id": "PA1004", "origin": "Toronto", "destination": "Hogsmeade",
     "departure_date": "2026-09-21", "days_until_departure": 7,
     "base_fare": 110.0, "seats_remaining": 60, "capacity": 100, "route_popularity": 0.80},
    {"flight_id": "PA1005", "origin": "London", "destination": "Paris",
     "departure_date": "2026-10-05", "days_until_departure": 21,
     "base_fare": 150.0, "seats_remaining": 100, "capacity": 150, "route_popularity": 0.85},
]

print("Number of flights:", len(flights))
print("First destination:", flights[0]["destination"])


Number of flights: 5
First destination: Hogsmeade


## Part B — Print and filter


In [4]:
for flight in flights:
    print(flight["flight_id"], flight["origin"], "to", flight["destination"],
          "| Base fare:", flight["base_fare"])


PA1001 London to Hogsmeade | Base fare: 135.0
PA1002 London to Edinburgh | Base fare: 92.0
PA1003 Manchester to Diagon Alley | Base fare: 96.0
PA1004 Toronto to Hogsmeade | Base fare: 110.0
PA1005 London to Paris | Base fare: 150.0


### Solution — filter by destination


In [5]:
chosen_destination = "Hogsmeade"

# SOLUTION
for flight in flights:
    if flight["destination"] == chosen_destination:
        print("Destination match:", flight["flight_id"], flight["destination"])


Destination match: PA1001 Hogsmeade
Destination match: PA1004 Hogsmeade


**Expected:** PA1001 and PA1004.


### Solution — filter by base-fare budget


In [6]:
budget = 110

# SOLUTION
for flight in flights:
    if flight["base_fare"] <= budget:
        print("Within budget:", flight["flight_id"], flight["destination"], flight["base_fare"])


Within budget: PA1002 Edinburgh 92.0
Within budget: PA1003 Diagon Alley 96.0
Within budget: PA1004 Hogsmeade 110.0


**Expected:** PA1002, PA1003, and PA1004.


## Part C — Pricing function


In [7]:
def calculate_fare(base_fare, days_until_departure, seats_remaining, capacity, route_popularity):
    # These two checks are provided.
    if capacity <= 0:
        raise ValueError("capacity must be greater than 0")
    if seats_remaining < 0 or seats_remaining > capacity:
        raise ValueError("seats_remaining must be between 0 and capacity")

    # SOLUTION 1
    if days_until_departure <= 7:
        time_factor = 1.35
    if 8 <=days_until_departure <= 21:
        time_factor = 1.10
    else:
        time_factor = 1.00

    # SOLUTION 2
    load_factor = 1 - seats_remaining / capacity
    capacity_factor = 1 + 0.45 * load_factor
    demand_factor = 0.9 + 0.3 * route_popularity

    # SOLUTION 3
    fare = base_fare * time_factor * capacity_factor * demand_factor
    if fare < 45:
        fare = 45
    elif fare > 1500:
        fare = 1500
    return round(fare, 2)


### Apply the function to all five flights


In [8]:
for flight in flights:
    # SOLUTION
    flight["fare"] = calculate_fare(
        flight["base_fare"],
        flight["days_until_departure"],
        flight["seats_remaining"],
        flight["capacity"],
        flight["route_popularity"],
    )
    print(flight["flight_id"], flight["destination"], flight["fare"])


PA1001 Hogsmeade 240.82
PA1002 Edinburgh 121.92
PA1003 Diagon Alley 149.84
PA1004 Hogsmeade 147.97
PA1005 Paris 219.16


**Expected fares:** 240.82, 121.92, 202.29, 199.76, 219.16.


### Try it — only five seats remaining


In [9]:
# SOLUTION
new_fare = calculate_fare(135, 12, 5, 120, 0.91)

print("Original fare:", flight["fare"])
print("Fare with 5 seats remaining:", new_fare)


Original fare: 219.16
Fare with 5 seats remaining: 249.31


**Expected:** the fare increases to 249.31.


## Optional bonus


In [10]:
destination = "Hogsmeade"
maximum_fare = 220

found = False
for flight in flights:
    if (flight["destination"] == destination
            and flight["fare"] <= maximum_fare
            and flight["seats_remaining"] > 0):
        print(flight["flight_id"], flight["destination"], flight["fare"])
        found = True

if not found:
    print("No matching flights.")


PA1004 Hogsmeade 147.97
